File: java/src/main/java/S017.java
import com.google.common.base.CharMatcher;
import com.google.common.base.Joiner;
import com.google.common.collect.ImmutableMap;
import com.google.common.collect.Lists;

import java.util.List;
import java.util.Map;
import java.util.stream.IntStream;

public class S017 {

    private static final Map<Integer, String> VALUE_TO_WORD = ImmutableMap.<Integer, String>builder()
            .put(1, "one")
            .put(2, "two")
            .put(3, "three")
            .put(4, "four")
            .put(5, "five")
            .put(6, "six")
            .put(7, "seven")
            .put(8, "eight")
            .put(9, "nine")
            .put(10, "ten")
            .put(11, "eleven")
            .put(12, "twelve")
            .put(13, "thirteen")
            .put(14, "fourteen")
            .put(15, "fifteen")
            .put(16, "sixteen")
            .put(17, "seventeen")
            .put(18, "eighteen")
            .put(19, "nineteen")
            .put(20, "twenty")
            .put(30, "thirty")
            .put(40, "forty")
            .put(50, "fifty")
            .put(60, "sixty")
            .put(70, "seventy")
            .put(80, "eighty")
            .put(90, "ninety")
            .put(100, "one hundred")
            .put(200, "two hundred")
            .put(300, "three hundred")
            .put(400, "four hundred")
            .put(500, "five hundred")
            .put(600, "six hundred")
            .put(700, "seven hundred")
            .put(800, "eight hundred")
            .put(900, "nine hundred")
            .put(1000, "one thousand")
            .build();

    public static int solve() {
        return IntStream.range(1, 1001).map(S017::letterCount).sum();
    }

    private static int letterCount(int n) {
        int remaining = n;
        List<String> words = Lists.newArrayList();
        while (remaining > 0) {
            int denom = getHighestDenom(remaining);
            words.add(VALUE_TO_WORD.get(denom));
            remaining -= denom;
        }
        String word = Joiner.on(' ').join(words);
        int length = CharMatcher.whitespace().removeFrom(word.toString()).length();
        if (n > 100 && n < 1000 && n % 100 != 0) {
            length += 3;
        }
        return length;
   }

    private static int getHighestDenom(int n) {
        if (n >= 1000) {
            return 1000;
        } else if (n >= 100) {
            return (n / 100) * 100;
        } else if (n >= 20) {
            return (n / 10) * 10;
        } else {
            return n;
        }
    }
}

File: cpp/S017.cpp
#include <iostream>
#include <vector>

using namespace std;

static const int LIMIT = 1000;

inline void add_word_length(vector<pair<int, int> >& vec, int number, string word) {
    vec.push_back(make_pair(number, word.length()));
}

vector<pair<int, int> > get_word_lengths() {
    vector<pair<int, int> > word_lengths;
    add_word_length(word_lengths, 1000, string("thousand"));
    add_word_length(word_lengths, 100, string("hundred"));
    add_word_length(word_lengths, 90, string("ninety"));
    add_word_length(word_lengths, 80, string("eighty"));
    add_word_length(word_lengths, 70, string("seventy"));
    add_word_length(word_lengths, 60, string("sixty"));
    add_word_length(word_lengths, 50, string("fifty"));
    add_word_length(word_lengths, 40, string("forty"));
    add_word_length(word_lengths, 30, string("thirty"));
    add_word_length(word_lengths, 20, string("twenty"));
    add_word_length(word_lengths, 19, string("nineteen"));
    add_word_length(word_lengths, 18, string("eighteen"));
    add_word_length(word_lengths, 17, string("seventeen"));
    add_word_length(word_lengths, 16, string("sixteen"));
    add_word_length(word_lengths, 15, string("fifteen"));
    add_word_length(word_lengths, 14, string("fourteen"));
    add_word_length(word_lengths, 13, string("thirteen"));
    add_word_length(word_lengths, 12, string("twelve"));
    add_word_length(word_lengths, 11, string("eleven"));
    add_word_length(word_lengths, 10, string("ten"));
    add_word_length(word_lengths, 9, string("nine"));
    add_word_length(word_lengths, 8, string("eight"));
    add_word_length(word_lengths, 7, string("seven"));
    add_word_length(word_lengths, 6, string("six"));
    add_word_length(word_lengths, 5, string("five"));
    add_word_length(word_lengths, 4, string("four"));
    add_word_length(word_lengths, 3, string("three"));
    add_word_length(word_lengths, 2, string("two"));
    add_word_length(word_lengths, 1, string("one"));
    return word_lengths;
}

int compute_word_length(vector<pair<int, int> > lengths, int n) {
    int remaining = n;
    int table_idx = 0;
    int total_length = 0;
    while (remaining > 0) {
        while (lengths[table_idx].first > remaining) {
            table_idx++;
        }
        int amount = lengths[table_idx].first;
        int length = lengths[table_idx].second;
        int scalar = remaining / amount;
        // scalar -> "one", "two", ...
        int scalar_length = amount >= 100 ? compute_word_length(lengths, scalar) : 0;
        total_length += scalar_length + length;
        remaining -= scalar * amount;
    }
    if (n > 100 && n % 100 != 0) {
        total_length += 3; // "and"
    }
    return total_length;
}

int main() {
    int sum = 0;
    vector<pair<int, int> > lengths = get_word_lengths();
    for (int i = 1; i <= LIMIT; i++) {
        sum += compute_word_length(lengths, i);
    }
    cout << sum << endl;
    return 0;
}
